src/_jaxbind.cc

```diff
@@ -61,67 +63,73 @@ template <typename T>
+  CNpArr py_kwargs = make_CArr_wrapper(dtp_kwargs, kwargs.untyped_data(), shape_kwargs);
+
+  auto n_args = args.size();
+  auto n_out = results.size();
 
-  nb::handle hnd(*reinterpret_cast<PyObject **>(in[0]));
-  auto func = nb::borrow<nb::object>(hnd);
 
-  size_t idx = 1;
-  size_t nargs = *reinterpret_cast<uint64_t *>(in[idx++]);
   nb::list py_in;
-  for (size_t i=0; i<nargs; i++) {
-    // Getting type, rank, and shape of the input
-    auto dtp_a = tcdict.at(uint8_t(*reinterpret_cast<int64_t *>(in[idx++])));
-    size_t ndim_a = *reinterpret_cast<uint64_t *>(in[idx++]);
+  for (size_t i=0; i<n_args; i++)
+    {
+    auto arg = args.get<ffi::AnyBuffer>(i).value();
+    auto dtp_a = tcdict.at(arg.element_type());
+    auto dims = arg.dimensions();
     shape_t shape_a;
-    for (size_t j=0; j<ndim_a; ++j) {
-      shape_a.push_back(*reinterpret_cast<uint64_t *>(in[idx++]));
-    }
+    for (auto x : dims) shape_a.push_back(x);
     // Building "pseudo" numpy arrays on top of the provided memory regions.
     // This should be completely fine, as long as the called function does not
     // keep any references to them.
-    CNpArr py_a = make_CArr_wrapper(dtp_a, in[idx++], shape_a);
+    CNpArr py_a = make_CArr_wrapper(dtp_a, arg.untyped_data(), shape_a);
     py_in.append(py_a);
-  }
+    }
 
-  // if we have only one output, out_raw is a void * pointing to the data of this output
-  // otherwise, out_raw is a void ** pointing to an array of void * pointing to the individual data
-  void **out = reinterpret_cast<void **>(out_raw);
-  void *out_single = reinterpret_cast<void *>(out_raw);
-  size_t nout = *reinterpret_cast<uint64_t *>(in[idx++]);
   nb::list py_out;
-  for (size_t i=0; i<nout; i++) {
-    // Getting type, rank, and shape of the output
-    auto dtp_out = tcdict.at(uint8_t(*reinterpret_cast<int64_t *>(in[idx++])));
-    size_t ndim_out = *reinterpret_cast<uint64_t *>(in[idx++]);
+  for (size_t i=0; i<n_out; i++) {
+    auto out = results.get<ffi::AnyBuffer>(i).value();
+    auto dtp_out = tcdict.at(out->element_type());
+    auto dims = out->dimensions();
     shape_t shape_out;
-    for (size_t j=0; j<ndim_out; ++j) {
-      shape_out.push_back(*reinterpret_cast<uint64_t *>(in[idx++]));
-    }
-    NpArr py_o = make_Arr_wrapper(dtp_out, (nout==1) ? out_single : out[i], shape_out);
+    for (auto x : dims) shape_out.push_back(x);
+    NpArr py_o = make_Arr_wrapper(dtp_out, out->untyped_data(), shape_out);
     py_out.append(py_o);
   }
 
-  auto dtp_kwargs = tcdict.at(uint8_t(*reinterpret_cast<int64_t *>(in[idx++])));
-  size_t size_kwargs = *reinterpret_cast<uint64_t *>(in[idx++]);
-  CNpArr py_kwargs = make_CArr_wrapper(dtp_kwargs, in[idx++], {size_kwargs});
-
-  // Execute the Python function implementing the desired operation
+  PyObject* raw_ptr = reinterpret_cast<PyObject*>(func_id);
+  nb::handle hnd(raw_ptr);
+  nb::object func = nb::borrow<nb::object>(hnd);
   func(py_out, py_in, py_kwargs);
```

Comment:
I think that errors in the called functions need to be communicated to the caller via return codes with the new FFI.
So I think the two lines below probably need to be replaced with just
`return func(py_out, py_in, py_kwargs);`
Still, that doesn't catch the exceptions we may be throwing in the lines above. Perhaps we need to guard everything by `try ... catch`; I'll look into that.